File: 백준/Gold/9663. N－Queen/N－Queen.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int     can_go(int *arr, int x)
{
        for (int i = 0; i < x; i++)
        {
                
                if (arr[i] == arr[x] || abs(arr[i] - arr[x]) == abs(i - x))
                {
                        return (0);
                }
        }
        return (1);

}

void     dfs(int *arr, int x, int n, int *ans)
{
        if (x == n)
        {
                (*ans)++;
                return ;
        }
        for (int i = 0; i < n; i++)
        {
                arr[x] = i; 

                if (can_go(arr, x))
                {
                        dfs(arr, x+1, n, ans);
                }
        }

}

int     main(void)
{       
        int     n;
        int     *arr;
        int     ans;

        scanf("%d", &n);
        ans = 0;
        arr = (int *)malloc(sizeof (int) * n);
        dfs(arr, 0, n, &ans);
        printf("%d", ans);
        free(arr);
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys
input = sys.stdin.readline

def can_go(x):

    for y in range(x):
        if (arr[y] == arr[x] or abs(arr[x] - arr[y]) == abs(x - y)):
            return False
    return True

def dfs(x):
    global ans

    if (x == n):
        ans += 1
        return

    for y in range(n):
        arr[x] = y
        if can_go(x):
            dfs(x + 1)


n = int(input())
arr = [float('inf') for _ in range(n)]
ans = 0
dfs(0)
print(ans)
